staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go

```diff
@@ -129,9 +130,32 @@ func ValidateNoNewFinalizers(newFinalizers []string, oldFinalizers []string, fld
 
 // ValidateImmutableField validates the new value and the old value are deeply equal.
 func ValidateImmutableField(newVal, oldVal interface{}, fldPath *field.Path) field.ErrorList {
+	return ValidateImmutableFieldWithErr(newVal, oldVal, field.ErrorTypeInvalid, fldPath, newVal, FieldImmutableErrorMsg)
+}
+
+// ValidateImmutableFieldWithErr validates the new value and the old value are deeply equal.
+// It accepts errorType fldPath value & detail to construct an error  when the result is not equal, and ValidateImmutableFieldWithErr will append additional error information
+// in err.detail and return this error.
+func ValidateImmutableFieldWithErr(newVal, oldVal interface{}, errorType field.ErrorType, fldPath *field.Path, value interface{}, detail string) field.ErrorList {
```

Comment:
https://github.com/kubernetes/kubernetes/pull/113963#discussion_r1064706542 says that you shouldn't pass the errorType?

Reply:
But if we do not provide errorType, should we just return an Invalid error? If we do so, then a lot of unit tests which rely on `ValidateImmutableFieldWithErr` will have to be changed, and it will be a huge change, so I prefer to pass the errorType to minimize out changes. 

But if you insist, please let me know, and I will change it soon.

Reply:
> [#113963 (comment)](https://github.com/kubernetes/kubernetes/pull/113963#discussion_r1064706542) says that you shouldn't pass the errorType?

@apelisse So what should I do now?

Reply:
> [#113963 (comment)](https://github.com/kubernetes/kubernetes/pull/113963#discussion_r1064706542) says that you shouldn't pass the errorType?

@apelisse So what should I do now? remove the errorType and make ValidateImmutableFieldWithErr always return InvalidError?

Reply:
I think we have basically three options, from simpler to more complicated:
1. We always use the same `ValidateIMmutableField` method everywhere
2. We allow customizing the error message but we always use `field.Invalid`
3. We allow custom error type and error message 

Since we've decided not to do 3, and since not doing 3 means that we're going to change the error message to some extent, I would say we go for 1. and just make things simple by calling `ValidateImmutableField` everywhere, use `field.Invalid` everywhere and keep things super simple.

Reply:
Codes have been revised according to solution1, @liggitt @apelisse @pohly, what about this time?